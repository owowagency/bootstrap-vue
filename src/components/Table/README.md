# Table

```vue
<Table :fields="[Field]" :items="[Object]" multiSort hover />
```

## Props

| Prop name    | Description                                        | Type                            | Values | Default     |
| ------------ | -------------------------------------------------- | ------------------------------- | ------ | ----------- |
| fields       | The fields of the table                            | Field[]                         | -      | undefined   |
| items        | The items that are shown in the table              | Record&lt;string, unknown&gt;[] | -      | () =&gt; [] |
| multiSort    | Defines if multiple items can be sorted at once    | boolean                         | -      | false       |
| hover        | Defines if the `table-hover` class should be used  | boolean                         | -      | false       |
| onClick:row  | Defines if the `table-click` class should be used  | func                            | -      | undefined   |
| tbodyTrClass | The classes that will be forwarded to the tbody tr | string\|array\|object\|func     | -      | undefined   |
| theadTrClass | The classes that will be forwarded to the tbody tr | string\|array\|object           | -      | undefined   |

## Slots

| Name                           | Description                                 | Bindings                                                                                                                                                                                                                                          |
| ------------------------------ | ------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `header-${header.key}`         | Displays the header of the table            | **key** `string` - The key of the header<br/>**header** `object` - The header object<br/>                                                                                                                                                         |
| `header-${header.key}-content` | Displays the content of the table header    | **header** `object` - The header object<br/>                                                                                                                                                                                                      |
| top-row                        | Displays as the top row of the table        | **headers** `object[]` - The headers of the table                                                                                                                                                                                                 |
| no-items                       | Displays if there are no items in the table | **headers** `object[]` - The headers of the table                                                                                                                                                                                                 |
| row                            | Displays the row of the table               | **item** `object` - The item object<br/>**index** `int` - The index of the item                                                                                                                                                                   |
| header.key                     | Displays the content of the table row       | **key** `any` - The key of the header<br/><br/>**field** `object` - The header object<br/>**item** `object` - The item of the row<br/>**index** `int` - The index of the item of the row<br/>**value** `any` - The value that should be displayed |
| `${header.key}-content`        | Displays the content of the table row       | **field** `any` - The header object<br/>**item** `object` - The item of the row<br/>**index** `int` - The index of the item of the row<br/>**value** `any` - The value that should be displayed<br/>                                              |
| bottom-row                     | Displays as the bottom row of the table     | **headers** `object[]` - The headers of the table                                                                                                                                                                                                 |

## Events

| Event name | Properties                                                                             | Description                                   |
| ---------- | -------------------------------------------------------------------------------------- | --------------------------------------------- |
| click:row  | **item** `object` - The row that was clicked                                           | Fired when a row is clicked                   |
| hover:row  | **item** `object` - The row that was hovered<br/>**undefined** `unknown` - e The event | Fired when a row is hovered                   |
| sort       | **sorted** `object` - The sort object                                                  | Fired when the sorting of a column is changed |
