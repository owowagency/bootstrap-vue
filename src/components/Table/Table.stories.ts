import OwTable from '.';

export default {
    title: 'Content/Table',
    component: OwTable,
    argTypes: {
    },
};

const template = (args) => ({
    components: {OwTable},
    setup() {
        return {args};
    },
    template: '<OwTable v-bind="args" />',
});

export const Default = template.bind({});

Default.args = {
    fields: [
        {key: 'firstName', label: 'First name', sortable: true, sort: 'asc'},
        {key: 'lastName', label: 'Last name', sortable: true},
        {key: 'gender', label: 'Gender'},
    ],
    items: [
        {firstName: 'Anakin', lastName: 'Skywalker', gender: 'Male'},
        {firstName: 'Jar Jar', lastName: 'Binks', gender: 'Female'},
    ],
    multiSort: false,
    hover: false,
};
