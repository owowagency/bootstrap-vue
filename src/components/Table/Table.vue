<template>
    <table
        class="table"
        :class="classes"
    >
        <thead v-if="headers.length">
            <tr>
                <slot
                    v-for="header in headers"
                    :key="`header-${header.key}`"
                    :header="header"
                    :name="`header-${header.key}`"
                >
                    <th
                        :key="header.key"
                        :class="{
                            'table-heading-sortable': header.sortable,
                            'table-heading-sortable-asc': sorted[header.key] === 'asc',
                            'table-heading-sortable-desc': sorted[header.key] === 'desc',
                        }"
                        @click="sort(header)"
                    >
                        {{ header.label || header.key }}
                    </th>
                </slot>
            </tr>
        </thead>

        <tbody>
            <slot name="top-row" />

            <slot
                v-if="!items.length"
                name="no-items"
            />

            <template
                v-for="(item, index) in items"
                v-else
                :key="`item-${index}`"
            >
                <slot
                    name="row"
                    :item="item"
                    :index="index"
                >
                    <tr @click="$emit('click:row', item)">
                        <slot
                            v-for="header in headers"
                            :key="header.key"
                            :name="header.key"
                            :item="item"
                            :index="index"
                        >
                            <td :key="`item-${index}-${header}`">
                                {{ item[header.key] }}
                            </td>
                        </slot>
                    </tr>
                </slot>
            </template>

            <slot name="bottom-row" />
        </tbody>
    </table>
</template>

<script lang="ts">
export interface Field {
    key: string;
    label?: string;
    sortable?: boolean;
    sort?: 'asc' | 'desc';
}
</script>

<script lang="ts" setup>
import {PropType, computed, ref} from 'vue';
import {useClasses} from '@/composables';

const props = defineProps({
    fields: {
        type: Array as PropType<Field[]>,
        required: false,
        default: undefined,
    },
    items: {
        type: Array as PropType<Record<string, unknown>[]>,
        default: () => [],
    },
    multiSort: {
        type: Boolean,
        default: false,
    },
    hover: {
        type: Boolean,
        default: false,
    },
    // $attrs doesn't return emitted events so
    // we have to catch it as a prop
    // eslint-disable-next-line vue/prop-name-casing
    'onClick:row': {
        type: Function,
        default: undefined,
    },
});

const emit = defineEmits<{
    (event: 'click:row', item: Record<string, unknown>): void
    (event: 'sort', sorted: Record<string, string>): void
}>();

const sorted = ref(
    (props.fields || [])
        .filter(f => f.sortable && f.sort)
        .reduce((obj, item) => Object.assign(obj, {[item.key]: item.sort}), {}),
);

const headers = computed<Field[]>(() => {
    if (props.fields) {
        return props.fields;
    }

    const first = props.items.find(i => i !== undefined);

    return first
        ? Object.keys(first)
            .map(f => ({key: f}))
        : [];
});

const sort = (field: Field) => {
    if (!field.sortable) {
        return;
    }

    const key = field.key;

    const sort = sorted.value[key];

    if (props.multiSort) {
        if (sort === 'asc') {
            sorted.value[key] = 'desc';
        } else if (sort === 'desc') {
            delete sorted.value[key];
        } else {
            sorted.value[key] = 'asc';
        }
    } else {
        sorted.value = {
            [key]: sort === 'asc' ? 'desc' : 'asc',
        };
    }

    emit('sort', sorted.value);
};

const {classes} = useClasses(computed(() => [
    props.hover && 'table-hover',
    props['onClick:row'] && 'table-click',
]));
</script>
