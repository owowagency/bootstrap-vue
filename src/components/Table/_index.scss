.table {
    .table-heading-sortable {
        cursor: pointer;

        &::after {
            vertical-align: middle;
            display: inline-block;
            content: '';
            height: 1em;
            margin: 0 .5rem;
            width: 1em;
        }

        &.table-heading-sortable-asc::after {
            background-image: escape-svg($table-sort-icon);
        }

        &.table-heading-sortable-desc::after {
            @extend .table-heading-sortable-asc;
            transform: rotate(180deg);
        }
    }
}

.table-click {
    tbody > tr {
        cursor: pointer;
    }
}
