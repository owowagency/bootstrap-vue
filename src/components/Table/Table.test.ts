import {describe, expect, it} from 'vitest';
import Table from '.';
import {shallowMount} from '@vue/test-utils';

const props = {
    items: [
        {id: 1, name: 'Word'},
        {id: 2, name: 'Press'},
        {id: 3, name: 'ISDABEST'},
        {id: 4, name: '#prank'},
    ],
};

const fields = [
    {label: 'ajdie', key: 'id', sortable: true, sort: 'asc'},
    {label: 'neem', key: 'name', sortable: true},
    {label: 'type', key: 'type'},
];

describe('template', () => {
    componentRenderTest(Table);

    componentRenderTest(Table, {props}, true, 'renders with items');

    componentRenderTest(
        Table,
        {
            props: {
                ...props,
                fields,
            },
        },
        true,
        'renders with fields and items',
    );

    componentSlotRenderTest(Table, 'header-id', {props});

    componentSlotRenderTest(Table, 'header-id-content', {props});

    componentSlotRenderTest(Table, 'row', {props});

    componentSlotRenderTest(Table, 'top-row', {props});

    componentSlotRenderTest(Table, 'bottom-row', {props});

    componentSlotRenderTest(Table, 'id', {props});

    componentSlotRenderTest(Table, 'id-content', {props});

    componentSlotRenderTest(Table, 'no-items', {
        props: {
            items: [],
        },
    });

    componentWrapperClassTest(Table, {hover: true}, 'table-hover');

    it('emits event on click table row', async() => {
        const wrapper = shallowMount(Table, {
            props: {
                ...props,
                'onClick:row': () => 'test',
            },
        });

        expect(wrapper.classes('table-click')).toBe(true);

        await wrapper.find('tbody tr').trigger('click');

        expect(wrapper.emitted('click:row')[0]).toEqual([props.items[0]]);
    });

    it('emits event on hover table row', async() => {
        const wrapper = shallowMount(Table, {
            props: {
                ...props,
                'onHover:row': () => 'test',
            },
        });

        await wrapper.find('tbody tr').trigger('mouseenter');

        expect(wrapper.emitted('hover:row')[0]).toEqual([props.items[0], expect.any(MouseEvent)]);
    });

    it('renders labelClass when given', () => {
        const wrapper = shallowMount(Table, {
            props: {
                ...props,
                theadTrClass: 'bg-gray-100',
                tbodyTrClass: 'bg-gray-200',
            },
        });

        const headTr = wrapper.find('thead tr');
        const bodyTr = wrapper.find('tbody tr');

        expect(headTr.classes()).toContain('bg-gray-100');
        expect(bodyTr.classes()).toContain('bg-gray-200');
    });

    it('renders tbody trClass when given as function', () => {
        const wrapper = shallowMount(Table, {
            props: {
                ...props,
                tbodyTrClass: (item: typeof props['items'][number]) => (item.id === 1 ? 'bg-gray-200' : 'bg-gray-100'),
            },
        });

        const bodyTr = wrapper.findAll('tbody tr');

        expect(bodyTr[0].classes()).toContain('bg-gray-200');
        expect(bodyTr[1].classes()).toContain('bg-gray-100');
    });
});

describe('headers', () => {
    it('is empty when no items nor fields have been given', () => {
        const wrapper = shallowMount(Table);

        expect(wrapper.vm.headers).toStrictEqual([]);
    });

    it('extracts from item when no fields given', () => {
        const wrapper = shallowMount(Table, {props});

        expect(wrapper.vm.headers).toStrictEqual([
            {key: 'id'},
            {key: 'name'},
        ]);
    });

    it('uses fields when given', () => {
        const wrapper = shallowMount(Table, {
            props: {fields},
        });

        expect(wrapper.vm.headers).toStrictEqual(fields);
    });
});

describe('sort', () => {
    it('sorts header if fields are sortable', async() => {
        const wrapper = shallowMount(Table, {
            props: {fields},
        });

        const th = wrapper.findAll('th');

        const firstTh = th[0];

        const secondTh = th[1];

        // Check if sorts are correct as start with defined properties, that is 'asc'
        expect(firstTh.classes()).toEqual(['table-heading-sortable', 'table-heading-sortable-asc']);

        expect(wrapper.vm.sorted).toStrictEqual({id: 'asc'});

        // After clicking first column heading sort directions should be 'desc'
        await firstTh.trigger('click');

        expect(firstTh.classes()).toEqual(['table-heading-sortable', 'table-heading-sortable-desc']);

        expect(wrapper.vm.sorted).toStrictEqual({id: 'desc'});

        expect(wrapper.emitted('sort')[0]).toEqual([{id: 'desc'}]);

        // After clicking second column heading, first heading should not
        // have sorting, and second column should have 'asc' sort
        await secondTh.trigger('click');

        expect(firstTh.classes()).toEqual(['table-heading-sortable']);

        expect(secondTh.classes()).toEqual(['table-heading-sortable', 'table-heading-sortable-asc']);

        expect(wrapper.vm.sorted).toStrictEqual({name: 'asc'});

        expect(wrapper.emitted('sort')[1]).toEqual([{name: 'asc'}]);
    });

    it('sorts multiple column on multi sort', async() => {
        const wrapper = shallowMount(Table, {
            props: {
                fields,
                multiSort: true,
            },
        });

        const th = wrapper.findAll('th');

        const firstTh = th[0];

        const secondTh = th[1];

        // Check if sorts are correct as start with defined properties, that is 'asc'
        expect(firstTh.classes()).toEqual(['table-heading-sortable', 'table-heading-sortable-asc']);

        expect(wrapper.vm.sorted).toStrictEqual({id: 'asc'});

        // After clicking first column heading sort directions should be 'desc'
        await firstTh.trigger('click');

        expect(firstTh.classes()).toEqual(['table-heading-sortable', 'table-heading-sortable-desc']);

        expect(wrapper.vm.sorted).toStrictEqual({id: 'desc'});

        expect(wrapper.emitted('sort')[0]).toEqual([{id: 'desc'}]);

        // After clicking second column, first column should stay untucked
        // and second column should have 'asc' sort
        await secondTh.trigger('click');

        expect(firstTh.classes()).toEqual(['table-heading-sortable', 'table-heading-sortable-desc']);

        expect(secondTh.classes()).toEqual(['table-heading-sortable', 'table-heading-sortable-asc']);

        expect(wrapper.vm.sorted).toStrictEqual({id: 'desc', name: 'asc'});

        expect(wrapper.emitted('sort')[1]).toEqual([{id: 'desc', name: 'asc'}]);

        // After clicking first column again, it should clear first column's sorting
        await firstTh.trigger('click');

        expect(firstTh.classes()).toEqual(['table-heading-sortable']);

        expect(secondTh.classes()).toEqual(['table-heading-sortable', 'table-heading-sortable-asc']);

        expect(wrapper.vm.sorted).toStrictEqual({name: 'asc'});

        expect(wrapper.emitted('sort')[2]).toEqual([{name: 'asc'}]);
    });

    it('does not emit when field is not sortable', async() => {
        const wrapper = shallowMount(Table, {
            props: {fields: [{label: 'yooo', key: 'oooy'}]},
        });

        const firstTh = wrapper.find('th');

        await firstTh.trigger('click');

        expect(wrapper.emitted('sort')).toBeUndefined();
    });
});
