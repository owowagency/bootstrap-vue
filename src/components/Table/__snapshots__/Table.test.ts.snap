// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `
"<table class=\\"table\\">
  <!--v-if-->
  <tbody></tbody>
</table>"
`;

exports[`template renders with fields and items 1`] = `
"<table class=\\"table\\">
  <thead>
    <tr>
      <th class=\\"table-heading-sortable table-heading-sortable-asc\\">ajdie</th>
      <th class=\\"table-heading-sortable\\">neem</th>
      <th class=\\"\\">type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>1</td>
      <td>Word</td>
      <td></td>
    </tr>
    <tr>
      <td>2</td>
      <td>Press</td>
      <td></td>
    </tr>
    <tr>
      <td>3</td>
      <td>ISDABEST</td>
      <td></td>
    </tr>
    <tr>
      <td>4</td>
      <td>#prank</td>
      <td></td>
    </tr>
  </tbody>
</table>"
`;

exports[`template renders with items 1`] = `
"<table class=\\"table\\">
  <thead>
    <tr>
      <th class=\\"\\">id</th>
      <th class=\\"\\">name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>1</td>
      <td>Word</td>
    </tr>
    <tr>
      <td>2</td>
      <td>Press</td>
    </tr>
    <tr>
      <td>3</td>
      <td>ISDABEST</td>
    </tr>
    <tr>
      <td>4</td>
      <td>#prank</td>
    </tr>
  </tbody>
</table>"
`;
